cicecore/cicedynB/dynamics/ice_dyn_evp.F90

```diff
@@ -393,81 +393,147 @@ subroutine evp (dt)
             divu,rdg_conv,rdg_shear,shear,taubx,tauby                     )
 
       else ! evp_algorithm == standard_2d (Standard CICE)
+         if (ndte > 1) then 
+            do ksub = 1,ndte-1       ! subcycling
```

Comment:
If ndte=1 then some compilers will create warnings and maybe even stop due to a do loop that runs from 1 to 0. The same should be included in the 1d .Now it aborts if ndte < 1

Reply:
It can't be true that the code will stop?  Compiler warnings maybe, but setting loop values so a loop won't be executed is done all the time and should be allowed in general.  I'm not opposed to the if check, it's not going to affect performance, but I think it's cleaner not being there.  Where is the check for ndte<1.  It's not in this part of the code, so the if check is needed even less.

Reply:
It seems a strong motivator is to get rid of if checks in loops.  I assume this is for performance.  My question continues to be whether we have measured a performance benefit with these changes.  If not, then I prefer that we not replicate the calls for the last subcycle step in order to add the deformations call outside an if statement.  My suspicion is that the if statement in the subcycling loop is not going to affect performance.  Getting rid of if statements inside an ij loop, ok maybe, but not at this high a level in the computations.  In my opinion, it's easier to maintain if the last iteration is not separated/duplicated.  Maybe there are other reasons for separating that last subcycle.  But if not, then I still prefer we not do it.

Reply:
I agree in principle with getting rid of if checks in loops.  They can make a big difference in performance, but it depends on the loop size and things like how long the pipeline is for alternative commands, which probably varies a good bit across hardware/compiler configurations.  Maybe it's gotten better with newer generations.  The other considerations here are not having duplicative code (which I also agree with) and making close-but-not-quite-duplicated code for various dynamics options as similar as possible.  Since the dynamics is typically the most expensive part of the sea ice simulation, and centers seem to be moving toward higher numbers of subcycles for better convergence, then optimizing these loops seems wise.  It's hard to judge, though, if the existing timing tests are inconclusive.  Question:  in the original code, the deformations routine is called at the end, rather than during the loop.  Is the argument for moving it up into the loop (or interrupting the loop) that the computed deformations need to be consistent with the velocities just prior to the last iteration?  Or is it not better that the deformations be computed from the final velocities?  (Maybe I'm not understanding something here?)